sqlx-core/src/rt/mod.rs

```diff
@@ -42,6 +42,29 @@ pub async fn timeout<F: Future>(duration: Duration, f: F) -> Result<F::Output, T
     missing_rt((duration, f))
 }
 
+pub async fn timeout_at<F: Future>(deadline: Instant, f: F) -> Result<F::Output, TimeoutError> {
+    #[cfg(feature = "_rt-tokio")]
+    if rt_tokio::available() {
+        return tokio::time::timeout_at(deadline.into(), f)
+            .await
+            .map_err(|_| TimeoutError(()));
+    }
+
+    #[cfg(feature = "_rt-async-std")]
+    {
+        let Some(duration) = deadline.checked_duration_since(Instant::now()) else {
+            return Err(TimeoutError(()));
+        };
+
+        async_std::future::timeout(duration, f)
+            .await
+            .map_err(|_| TimeoutError(()))
+    }
+
+    #[cfg(not(feature = "_rt-async-std"))]
+    missing_rt((duration, f))
```

Comment:
When playing around with this PR locally (to see if it fixes an acquire timeout issue, which it unfortunately doesn't), I found that this caused a compile error. I think it should be

```suggestion
    missing_rt((deadline, f))
```

Reply:
@jplatte if you have a solid repro for acquire timeouts, I'd love to add it as a test here.

Reply:
I wish. It's in the proprietary version of the main work codebase, and somehow only happens w/ hyper 1.0 / axum 0.7. But if other debugging approaches don't work out, I can try the hyper upgrade on the much smaller OSS version of the codebase and reduce from there next week.

Reply:
One thing that Axum does is cancel the handler future if the client disconnects. I wonder if it's triggering a cancellation bug somewhere.

Do you have a `before_acquire` callback set?

Reply:
I did some digging a few weeks back and realized that connections could potentially get stuck in `return_to_pool` because there's no timeout: https://github.com/estuary/flow/issues/1676#issuecomment-2392077544

That's a change I was meaning to add to this PR but hadn't gotten to yet.

Reply:
I don't think it's a cancelation bug. It happens in a test that does a bunch of requests in parallel (50 originally, I can turn it down to 20 and still reliably reproduce the hang, but at 18 it succeded).

Reply:
What's the max size of the test pool?

And what's the acquire timeout set at?

Reply:
Hmmm, the max size of the pool is exactly 20, and once I use that amount of parallelism it breaks. Tried 19 too and that works. Acquire timeout is 20s, much longer than it takes the test to run to completion with up to 19 parallel requests.

I also tried raising the pool size, exactly same thing: Once the number of parallel requests is at least as big as the pool size, boom.

Further, I was using a `tokio::sync::Barrier` and separate `reqwest::Client`s, tokio tasks for the requests to happen as closely together as possible (this test was originally written to catch another race). If I don't make the tasks wait on the barrier before making the request, that seems to already mix things up sufficiently for the test succeed, even at a pool size of 20 and 50 parallel requests.

Reply:
I found the bug, it has nothing to do with SQLx itself. The test was deadlocking the server in a really weird way (related to the DB pool).